File: Medium/752. Open the Lock/Solution.java
class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> used = new HashSet(Arrays.asList(deadends));
        Queue<String> queue = new LinkedList();
        int moves = -1;

        if (!used.contains("0000"))
            queue.offer("0000");

        while (!queue.isEmpty()) {
            moves++;
            int s = queue.size();
            for (int i = 0; i < s; i++) {
                String cur = queue.poll();
                if (cur.equals(target))
                    return moves;

                for (String next : adjacentMoves(cur)) {
                    if (!used.contains(next)) {
                        queue.offer(next);
                        used.add(next);
                    }
                }
            }
        }

        return -1;
    }

    private List<String> adjacentMoves(String state) {
        List<String> adj = new ArrayList();
        for (int i = 0; i < state.length(); i++) {
            adj.add(state.substring(0, i) + forward(state.charAt(i)) + state.substring(i + 1));
            adj.add(state.substring(0, i) + rewind(state.charAt(i)) + state.substring(i + 1));
        }

        return adj;
    }

    private int forward(int x) {
        x -= '0';
        return x == 9 ? 0 : (x + 1);
    }

    private int rewind(int x) {
        x -= '0';
        return x == 0 ? 9 : (x - 1);
    }
}

File: Medium/752. Open the Lock/solution.js
/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
var openLock = function (deadends, target) {
  let seen = new Set(deadends);
  let steps = -1;
  let start = "0000";
  if (seen.has(start)) return steps;

  let queue = [start];

  while (queue.length > 0) {
    steps++;

    let size = queue.length;
    for (let i = 0; i < size; i++) {
      let current = queue.shift();
      if (current == target) return steps;

      for (let next of adj(current)) {
        if (seen.has(next)) continue;

        seen.add(next);
        queue.push(next);
      }
    }
  }

  return -1;

  function adj(state) {
    ff = (s, i) => s.substring(0, i) + ((Number(s[i]) + 1) % 10) + s.substring(i + 1);
    rw = (s, i) => s.substring(0, i) + ((Number(s[i]) + 9) % 10) + s.substring(i + 1);
    let arr = Array.from(state);
    return [...arr.map((_,i) => ff(state, i)), ...arr.map((_,i) => rw(state, i))];
  }
};
